src/main/kotlin/uk/gov/dluhc/printapi/service/CertificateDataRetentionService.kt

```diff
@@ -0,0 +1,42 @@
+package uk.gov.dluhc.printapi.service
+
+import mu.KotlinLogging
+import org.springframework.stereotype.Service
+import uk.gov.dluhc.printapi.database.repository.CertificateRepository
+import uk.gov.dluhc.printapi.exception.CertificateNotFoundException
+import uk.gov.dluhc.printapi.mapper.SourceTypeMapper
+import uk.gov.dluhc.printapi.messaging.models.ApplicationRemovedMessage
+import javax.transaction.Transactional
+
+private val logger = KotlinLogging.logger {}
+
+@Service
+class CertificateDataRetentionService(
+    private val sourceTypeMapper: SourceTypeMapper,
+    private val certificateRepository: CertificateRepository,
+    private val certificateRemovalDateResolver: CertificateRemovalDateResolver
+) {
+    /**
+     * Sets the initialRetentionRemovalDate on a [uk.gov.dluhc.printapi.database.entity.Certificate], after the
+     * originating application is removed from the source system (e.g. VCA).
+     *
+     * @param message An [ApplicationRemovedMessage] sent from the source system.
+     */
+    @Transactional
+    fun handleSourceApplicationRemoved(message: ApplicationRemovedMessage) {
+
+        with(message) {
+            val sourceType = sourceTypeMapper.mapSqsToEntity(sourceType)
+            val certificate = certificateRepository.findByGssCodeAndSourceTypeAndSourceReference(
+                gssCode = gssCode,
+                sourceType = sourceType,
+                sourceReference = sourceReference
+            )?.also {
+                it.initialRetentionRemovalDate =
+                    certificateRemovalDateResolver.getCertificateInitialRetentionPeriodRemovalDate(it.issueDate, gssCode)
+            } ?: throw CertificateNotFoundException(sourceType, sourceReference).also { logger.warn(it.message) }
```

Comment:
We might want to think about the handling of this exception.
I think you're right to handle the eventuality that we might not find the certificate in the repository, but this service method is being called by a listener, which means (unless you catch and swallow it in the listener) the SQS message will fail, and the message will be re-queued and re-tried up to the retry limit (6 times ?), then DLQ'ed

If the certificate is not found in the repository, is there any milage in trying again? Do we think it might find the certificate in the database on a subsequent retry ?

Reply:
That's a good point - I'll change it to log an error. However, I expect (and sincerely hope) this never happens!